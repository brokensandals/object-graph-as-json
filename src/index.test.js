import { Encoder, UnsafeDecoder } from './';

test('sample usage', () => {
  // Sample object
  const counter = {
    current: 1,
    increment() {
      this.current++;
    }
  };
  counter.increment();

  expect(counter.current).toEqual(2);
  
  // Encoding
  const encoder = new Encoder();
  const encoded = encoder.encode(counter);
  expect(encoded).toMatchSnapshot();

  // The encoded object can be serialized and deserialized to JSON if desired
  const serialized = JSON.stringify(encoded);
  const deserialized = JSON.parse(serialized);

  // Decoding
  const decoder = new UnsafeDecoder();
  const decoded = decoder.decode(deserialized);
  expect(decoded.current).toEqual(2);
  decoded.increment();
  expect(decoded.current).toEqual(3);
});

test('one of the reasons UnsafeDecoder is unsafe', () => {
  expect(Object.EVIL_THING_HAPPENED).toBeUndefined();

  const maliciousInput = {
    type: 'function',
    id: '1',
    // Normally, a function's source would start with something like
    // function(). Because this doesn't, the malicious code gets execute
    // as soon as we try to decode the input, even if we never try
    // to call the decoded function.
    source: 'Object.EVIL_THING_HAPPENED = true'
  };

  new UnsafeDecoder().decode(maliciousInput);

  expect(Object.EVIL_THING_HAPPENED).toBeTruthy();
});
