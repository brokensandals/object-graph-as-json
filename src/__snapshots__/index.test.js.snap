// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sample usage 1`] = `
Object {
  ".current": 2,
  ".increment": Object {
    ".length": Object {
      "configurable": true,
      "type": "property",
      "value": 0,
    },
    ".name": Object {
      "configurable": true,
      "type": "property",
      "value": "increment",
    },
    ".prototype": Object {
      "type": "property",
      "value": Object {
        ".constructor": Object {
          "configurable": true,
          "type": "property",
          "value": Object {
            "id": "2",
            "type": "ref",
          },
          "writable": true,
        },
        "id": "3",
        "type": "object",
      },
      "writable": true,
    },
    "id": "2",
    "source": "function increment() {
      this.current++;
    }",
    "type": "function",
  },
  "id": "1",
  "type": "object",
}
`;
